### PR TITLE
Add doc page with providers deprecations

### PR DESCRIPTION
**Pull request by kacpermuda**
## TLDR
I want to add a documentation page that will have all the deprecated classes, methods, functions etc. It will get more information when #36876 is merged (where i switched all deprecations to decorators, so that i can detect them with static analysis).

## Rationale
This is a step number 3 of the process described in #36876 , please refer to it for the rationale and in depth description.

## Future possibilities
We can probably incorporate this information inside the actual class/method docs and not in separate page.

---
**^ Add meaningful description above**
Read the **[Pull Request Guidelines](https://github.com/apache/airflow/blob/main/contributing-docs/05_pull_requests.rst#pull-request-guidelines)** for more information.
In case of fundamental code changes, an Airflow Improvement Proposal ([AIP](https://cwiki.apache.org/confluence/display/AIRFLOW/Airflow+Improvement+Proposals)) is needed.
In case of a new dependency, check compliance with the [ASF 3rd Party License Policy](https://www.apache.org/legal/resolved.html#category-x).
In case of backwards incompatible changes please leave a note in a newsfragment file, named `{pr_number}.significant.rst` or `{issue_number}.significant.rst`, in [newsfragments](https://github.com/apache/airflow/tree/main/newsfragments).

### CODE REVIEW

File: docs/apache-airflow-providers/core-extensions/deprecations.rst [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+    "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+    KIND, either express or implied.  See the License for the
+    specific language governing permissions and limitations
+    under the License.
+
+
+Deprecations
+==========================================
+
+This is a summary of deprecated objects in Apache Airflow Providers Packages.
```
Comment (eladkal) [inline docs/apache-airflow-providers/core-extensions/deprecations.rst]: If I understand it right what we will show here is deprecations only on the latest version of each provider. If so we should mention this

File: docs/spelling_wordlist.txt [hunk 2]
```diff
@@ -122,6 +122,7 @@ avp
 Avro
 avro
 aws
+AwsBaseAsyncHook
```
Comment (eladkal) [inline docs/spelling_wordlist.txt]: why was this added?

File: docs/apache-airflow-providers/core-extensions/deprecations.rst [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+
+.. note::
+   At the moment we only show deprecations for classes, functions, methods, properties etc.
+   Support for an argument deprecation or an argument value deprecation will be added in the future.
+
+
+.. important::
+   Only deprecations made with @deprecated decorator are included in this docs.
+   If You use warnings.warn in function/method body it will not show up in this documentation.
+   Make sure to use @deprecated decorator as shown in the below example:
```
Comment (eladkal) [inline docs/apache-airflow-providers/core-extensions/deprecations.rst]: Who is this addressed to?
This sounds more like information for contributors the `deprecations.rst` is for users.

File: docs/spelling_wordlist.txt (hunk 2 above)
Comment (kacpermuda) [reply docs/spelling_wordlist.txt]: Ah, this will be needed when #36876 is merged (soon i hope), i cherry picked it for testing and got some spelling errors there when building docs.

File: docs/apache-airflow-providers/core-extensions/deprecations.rst (hunk 3 above)
Comment (kacpermuda) [reply docs/apache-airflow-providers/core-extensions/deprecations.rst]: True, this is mostly for contributors or developers that are trying to see if the deprecation from their package is visible. Do You have any idea where can i put it? Or just leave it as a comment maybe?

File: docs/apache-airflow-providers/core-extensions/deprecations.rst (hunk 1 above)
Comment (kacpermuda) [reply docs/apache-airflow-providers/core-extensions/deprecations.rst]: Yes, that works only for latest versions, i added that to the message.

File: docs/spelling_wordlist.txt (hunk 2 above)
Comment (eladkal) [reply docs/spelling_wordlist.txt]: That is odd
I don't think the right fix is to add keywords that you didn't add to the exclusion list.

I wonder if maybe this automated process adds new words to the doc build (which were not there before) thus it means that everytime we make a release we will need to handle more new words to the exclusion list?

File: docs/apache-airflow-providers/core-extensions/deprecations.rst (hunk 3 above)
Comment (eladkal) [reply docs/apache-airflow-providers/core-extensions/deprecations.rst]: It should be in some contributor facing doc but I think it's better to find a way to enforce it rather than documenting it (for example: pre-commit that verify that all deprecations must be in the method you set as expected)
For example: We added pre-commit that verifies in providers we use only `AirflowProviderDeprecationWarning` so we may need another pre-commit to verify the method this `AirflowProviderDeprecationWarning` is being used.

https://github.com/apache/airflow/blob/48158c9967c14acf65264e338b8dea60af3f25e1/.pre-commit-config.yaml#L517-L521

Comment (kacpermuda) [reply docs/apache-airflow-providers/core-extensions/deprecations.rst]: I agree with the idea of enforcement rather than documenting, but i think this one is tricky at the moment. I believe ideally we should verify if there are no `warnings.warn` with`AirflowProviderDeprecationWarning` being raised anywhere in the function/method body, only in decorators, but at the moment there is no decorator for deprecating argument or their values. So we can't really tell if the `warnings.warn` in `__init__` is about the class or some argument of init (without parsing the message and taking a guess). Do You have any idea how i could enforce it ?

File: docs/spelling_wordlist.txt (hunk 2 above)
Comment (kacpermuda) [reply docs/spelling_wordlist.txt]: In the doc page we are providing instructions on what a user should use instead of a deprecated entity, so it's usually a message with path or name of another Airflow class,method, function etc. So i think yes, it may happen that with every build there might be a new "unknown" word if there is a new deprecation and new class included in the deprecation message. I am not sure if there is a way to avoid it? Developers can put anything in the deprecation message and it may cause spelling checks to fail, but it should be caught in the PR that adds the deprecation.

File: docs/apache-airflow-providers/core-extensions/deprecations.rst (hunk 3 above)
Comment (kacpermuda) [reply docs/apache-airflow-providers/core-extensions/deprecations.rst]: @eladkal Do You have any thoughts on that maybe ?

Comment (potiuk) [reply docs/apache-airflow-providers/core-extensions/deprecations.rst]: Chiming in here. I think it's difficult to make the distinction now. but we should be able to influence the way deprecations are raised.  

For example we could define a new AirfllowProviderAttributeDeprecationWarning(AirflowProviderDeprecationWarning) - taking the name of attribute, and new_name of the atribute as parameters. And only allow to use those in the body of the method.

Comment (potiuk) [reply docs/apache-airflow-providers/core-extensions/deprecations.rst]: I think the documentation on deprecations targetted for contributors, should find its way to the new contributing docs - likely here: https://github.com/apache/airflow/blob/main/contributing-docs/05_pull_requests.rst#coding-style-and-best-practices

File: docs/spelling_wordlist.txt (hunk 2 above)
Comment (potiuk) [reply docs/spelling_wordlist.txt]: @kacpermuda -> in case of those things that were added, we usually do it differently - names of the classes are usually `BaseAsyncSessionFactory` quoted with backticks - to make their fonts fixed width, so that they look like a code. And that automatically disables spell-checking for those names. That's why our spellchecking wordlis is not cluttered with 100s of those.

Comment (kacpermuda) [reply docs/spelling_wordlist.txt]: @potiuk That makes sense, I will adjust the class names that were missing backticks, so that i don't have to whitelist them in spellchecking. 

That leaves me with the question: if we start including deprecation message in the docs, when a developer forgets to put those backticks in the deprecation message, the spellcheck will fail and it will be to the person accepting the PR to make sure that they do not add those words to the whitelist but put backticks around them? Is there any way to add a clear information to the developer about why the spellcheck failed in this scenario so that he can correct himself ?

Comment (potiuk) [reply docs/spelling_wordlist.txt]: Sure - we have docs, we can explain there, also there are messages printed after doc build runs -  we already print quite a lot of information there - for example how you can build docs faster with selecting individual packages. Maybe a good idea to open PR with such change. 

Docs Troubleshooting:
  
https://github.com/apache/airflow/blob/main/docs/README.rst#docs-troubleshooting 


Hints (we print them coloured, to make them stand out): 

 https://github.com/apache/airflow/blob/main/docs/build_docs.py#L81

the place where they are printed is here:

 https://github.com/apache/airflow/blob/9f4f208b5da38bc2e82db682c636ec4fcf7ad617/docs/build_docs.py#L554  

You could check if there are spelling errors and give some hints to the user if they do. That would be a nice improvement  - probably you are the right person to describe it in the way that will help others like you, who had this situation :)

File: docs/apache-airflow-providers/core-extensions/deprecations.rst (hunk 3 above)
Comment (kacpermuda) [reply docs/apache-airflow-providers/core-extensions/deprecations.rst]: @potiuk Yes, i believe we can standardize the way the deprecations are made. I like the idea of specific deprecation warnings, but, correct me if I'm wrong, if we put this new kind of Warning in the core Airflow ( f.e. `airflow/exceptions.py`) then the providers will not be able to use it unless they set the required minimum version of Airflow to f.e. 2.8.2 (the next one from now), which can never happen for some of them, therefore, it's going to be hard to rely on it in documentation? We can ofc try some try/except on import, but I'm not sure if it will solve the problem ?

My idea for standarization of deprecations is to use decorators #36952 . We could define decorators like: @deprecated, @deprecated_args, @deprecated_arg_values like in tensorflow [codebase](https://github.com/tensorflow/tensorflow/blob/dec8e0b11f4f87693b67e125e67dfbc68d26c205/tensorflow/python/util/deprecation.py#L274), we could decorate the exact function/class it touches and possibly eliminate the need of using warnings.warn within the body.

Anyway, maybe it will be good for this particular PR to remove this part of docs that we refer to here, and mark this doc as experimental or something similar? Then we can try to work on some rules, document (in the place You mentioned) and enforce them in a separate PR, f.e. the one i mentioned. Let me know what You think.

Comment (potiuk) [reply docs/apache-airflow-providers/core-extensions/deprecations.rst]: Yes. You are right it's tricky for providers to use new features imported from common airlfow code. 
I thought about it and there is a way, though rather complex-ish - we could release another common provider  (`common.utils` for the lack of better name though I hate generic utils) where we could put such `never-backwards-incompatible` base classes that could be imported by providers. Similar to common.io. 

That would remove the dependency on airflow core and your https://github.com/apache/airflow/pull/36952 could land there, rather than in airflow core (or at least the part of it that is going to be used in providers).

And this is probably the only way we can implement decorators that we will be able to use in providers immediately (assuming that the `common.utils` will never be incompatible.

Maybe that's the right approach ?

Comment (kacpermuda) [reply docs/apache-airflow-providers/core-extensions/deprecations.rst]: I like the idea. Maybe there are (or will be) some other ideas that we'll want to implement as "globally available" without depending on core airflow version. Maybe let's move the discussion to #36952, and right here i will simply remove this paragraph from the docs for now.

File: docs/spelling_wordlist.txt (hunk 2 above)
Comment (kacpermuda) [reply docs/spelling_wordlist.txt]: I added missing backticks and also adjusted the printing of the spelling error, so the misspelled word stand out. I also added a paragraph to the docs in case somebody else encounter the same error and tries to add class names to  `docs/spelling_wordlist.txt`

File: docs/apache-airflow-providers/core-extensions/deprecations.rst (hunk 3 above)
Comment (kacpermuda) [reply docs/apache-airflow-providers/core-extensions/deprecations.rst]: As mentioned, i removed the part of the docs about how the deprecation should be made - it will be enforced once we implement a standardized way of deprecating across airflow codebase.

**Review by potiuk — APPROVED** [review]

File: docs/spelling_wordlist.txt (hunk 2 above)
Comment (eladkal) [reply docs/spelling_wordlist.txt]: Nice!

File: docs/exts/templates/deprecations.rst.jinja2 [hunk 4]
```diff
@@ -0,0 +1,37 @@
+{#
```
Comment (eladkal) [inline docs/exts/templates/deprecations.rst.jinja2:1]: Question...
is it possible to also have a version of this template per provider in the specific provider package?
so that Amazon deprecations will be shown in the Amazon provider directly
(Maybe is a future enhancement in a followup PR?)

**Review by eladkal — APPROVED** [review]

File: docs/exts/templates/deprecations.rst.jinja2 (hunk 4 above)
Comment (potiuk) [reply docs/exts/templates/deprecations.rst.jinja2:1]: Yeah.  This looks like a follow up :)

Comment (kacpermuda) [reply docs/exts/templates/deprecations.rst.jinja2:1]: I think it's a great idea. For me, the closer we get to the actual code docs with the deprecation information, the better :)

**Review by mobuchowski — APPROVED** [review]

**Comment by eladkal** [comment]
Thank you @kacpermuda for this one!
It should really help users to migrate faster to new versions
